cpp/src/binaryop/compiled/operation.cuh

```diff
@@ -0,0 +1,309 @@
+      col.element<Element>(i) = static_cast<Element>(val);
+    }
+  }
+};
+
+// TODO use inheritance. (or some other way to simplify the operator.)
+struct Add {
+  template <typename TypeCommon>
+  static constexpr inline bool is_supported()
+  {
+    return cudf::binops::compiled::CHECK::PlusExists<TypeCommon, TypeCommon>::value;
+  }
+
+  template <typename TypeLhs, typename TypeRhs>
+  static constexpr inline bool is_supported()
+  {
+    return !has_common_type_v<TypeLhs, TypeRhs> and
+           cudf::binops::compiled::CHECK::PlusExists<TypeLhs, TypeRhs>::value;
+    //(is_chrono<TypeLhs>() and is_chrono<TypeRhs>()) and
+    //!(is_timestamp<TypeLhs>() and is_timestamp<TypeRhs>());
+  }
+
+  // 1. With common type. (single dispatch, + typecast dispatch)
+  // Allow sum between chronos only when both input and output types
+  // are chronos. Unsupported combinations will fail to compile
+  template <typename TypeCommon, std::enable_if_t<is_supported<TypeCommon>()>* = nullptr>
+  CUDA_HOST_DEVICE_CALLABLE void operator()(size_type i,
+                                            column_device_view const lhs,
+                                            column_device_view const rhs,
+                                            mutable_column_device_view const out)
+  {
+    TypeCommon x = type_dispatcher(lhs.type(), type_casted_accessor<TypeCommon>{}, i, lhs);
+    TypeCommon y = type_dispatcher(rhs.type(), type_casted_accessor<TypeCommon>{}, i, rhs);
+    auto result  = x + y;
+    type_dispatcher(out.type(), typed_casted_writer<decltype(result)>{}, i, out, result);
+  }
+
+  // 2. without common type. (double dispatch, + out typecast dispatch)
+
+  // chronos t+d, d+t, d+d, !(t+t)
+  template <typename TypeLhs,
+            typename TypeRhs,
+            std::enable_if_t<is_supported<TypeLhs, TypeRhs>()
+                             //(is_chrono<TypeLhs>() and is_chrono<TypeRhs>()) and
+                             //!(is_timestamp<TypeLhs>() and is_timestamp<TypeRhs>())
+                             >* = nullptr>
+  CUDA_HOST_DEVICE_CALLABLE void operator()(size_type i,
+                                            column_device_view const lhs,
+                                            column_device_view const rhs,
+                                            mutable_column_device_view const out)
+  {
+    TypeLhs x   = lhs.element<TypeLhs>(i);
+    TypeRhs y   = rhs.element<TypeRhs>(i);
+    auto result = x + y;
+    type_dispatcher(out.type(), typed_casted_writer<decltype(result)>{}, i, out, result);
+  }
+};
+
+struct Sub {
+  template <typename TypeCommon>
+  static constexpr inline bool is_supported()
+  {
+    return cudf::binops::compiled::CHECK::MinusExists<TypeCommon, TypeCommon>::value;
+  }
+
+  template <typename TypeLhs, typename TypeRhs>
+  static constexpr inline bool is_supported()
+  {
+    return (!has_common_type_v<TypeLhs, TypeRhs>)and cudf::binops::compiled::CHECK::
+      MinusExists<TypeLhs, TypeRhs>::value;
+  }
```

Comment:
This can be significantly simplified by instead using `is_invocable` for the callsite SFINAE.

Reply:
Here, problem is not invocation with those arguments. It's about existence of that template function (whether that template function is able to compile or not). 

Already tried this. 
now, tried `std::is_invocable_v<decltype(&Sub::operate<TypeLhs, TypeRhs>), TypeLhs, TypeRhs>`. I get errors like
```
../../../../src/binaryop/compiled/operation.cuh(112):   error : no operator "-" matches these operands
            operand types are: __nv_bool - cudf::string_view
          detected during:
            instantiation of "auto cudf::binops::compiled::Sub::operate(TypeLhs, TypeRhs) [with TypeLhs=__nv_bool, TypeRhs=cudf::string_view]" 
(126): here```

Reply:
This works. Nice. Instead of static function, member function operator() is used for `is_invocable`.

But compile time shot up.
 Old code    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; : Add, Sub, Mul, Div, TrueDiv, FloorDiv =  58s
`is_invocable`: Add, Sub, Mul, Div, TrueDiv, FloorDiv = 329s

If I add single tparam `operator()<T>(T, T)`, compile time reduces to 222s. but still higher than old code.

single ops compilation time are 27s(Add), 29s(Sub), 33s(Mul), 46s(Div).

Reply:
Split to multiple files. `is_invocable` used.